fairlearn/reductions/_moments/cdf_demographic_parity_moment.py

```diff
@@ -0,0 +1,198 @@
+        the more general case where the grids are not equally spaced.
+
+        The binary label for each new data point is determined by whether the 
+        corresponding weight is positve or not.
+
+
+        Parameters
+        ----------
+        X : numpy.ndarray or pandas.DataFrame
+            Feature data
+        y : numpy.ndarray, pandas.DataFrame, pandas.Series, or list
+            Label vector
+
+        
+        Returns
+        -------
+        pandas.DataFrame, pandas.DataFrame, pandas.DataFrame, pandas.DataFrame
+            Represent the augmented features, sensitive features, labels, and weights.
+        """
+        self.n = np.shape(x)[0]
+        self.width = self.grids[1] - self.grids[0]
+
+        X_aug = pd.concat(repeat(x, self.grid_num))
+        A_aug = pd.concat(repeat(a, self.grid_num))
+        Y_values = pd.concat(repeat(y, self.grid_num))
+
+        theta_list = [s for theta in self.grids for s in repeat(theta, self.n)]
+        X_aug['theta'] = pd.Series(theta_list, index=X_aug.index)
+
+        X_aug.index = range(self.n * self.grid_num)
+        A_aug.index = range(self.n * self.grid_num)
+        Y_values.index = range(self.n * self.grid_num)
+
+        def weight_assign(theta, y):
+            return (self.loss.eval([self.next_grid(_theta) for _theta in theta], y) -
+                    self.loss.eval([self.prev_grid(_theta) for _theta in theta], y))
+        W = weight_assign(X_aug['theta'], Y_values)
+        Y_aug = 1*(W < 0)
+        W = abs(W)
+        return X_aug, A_aug, Y_aug, W
+
+    def load_data(self, X, y, **kwargs):
+        super().load_data(X, y, **kwargs)
+        # augment data
+        X_aug, A_aug, Y_aug, W = self.augment_data(self.X, self.tags[_GROUP_ID], self.tags[_LABEL])
+        self.objective.load_augmented_data(X_aug, Y_aug, sensitive_features=A_aug, weights=W)
+
+        # pass augmented data to utility_parity
+        self.utility_parity.load_data(X_aug, Y_aug, event=X_aug["theta"], sensitive_features=A_aug)
+        self.index = self.utility_parity.index
+
+        empirical_y_range = (self.tags[_LABEL].min(), self.tags[_LABEL].max())
+        assert (self.y_range[0] <= empirical_y_range[0])\
+            and (self.y_range[1] >= empirical_y_range[1]),\
+            "The actual range of the label y exceeds that required!"
+
+    def default_objective(self):
+        return self.objective
+
+    def bound(self):
+        return self.utility_parity.bound()
+
+    def project_lambda(self, lambda_vec):
+        return self.utility_parity.project_lambda(lambda_vec)
+
+    def gamma(self, predictor):
+        def classifier(X):
+            x = X.drop(["theta"], axis=1).iloc[:self.n]
+            pred = pd.Series(predictor(x))
+            pred_binary = pd.concat(repeat(pred, self.grid_num))
+            pred_binary.index = range(self.n * self.grid_num)
+            return 1*(pred_binary - X['theta'] >= 0)
```

Comment:
Are you trying to get `(pred_binary - X['theta']) >= 0` or `pred_binary - (X['theta'] >= 0)`? Probably better to be explicit even though operator precedence is clear.